File: bench/shapes.cpp
#include <vector>
#include <algorithm>
#include <string>
#include <iostream>
#include <sstream>
#include <stdio.h>

int rand(int n) { return rand() % n; }

class Shape {
public: std::string name;
public: Shape(const std::string &_name) : name(_name) { }
public: virtual void draw() { std::cout << "drawing: " << name << std::endl; }
};

class Circle: public Shape {
public: double radius;
public: Circle(const std::string &_name, int _radius)
  : Shape(_name), radius(_radius) {}

public: virtual void draw() {
  std::cout << "drawing circle: " << name << " r=" << radius << std::endl;
}
};

class Square: public Shape {
public: double side;
public: Square(const std::string &_name, int _side)
  : Shape(_name), side(_side) {}
public: void draw() {
  std::cout << "drawing square: " << name << " s=" << side << std::endl;  
}
};

class Shapes: public Shape {
public: Shapes() : Shape("shapes") {}
public: std::vector < std::shared_ptr < Shape > > parts;
public: void make(int n) {
  char cbuf[1024];
  parts.resize(n, std::shared_ptr < Shape > (0));
  for (int i=0; i<n; ++i) {
    if ((i%10) <= 5) {
      snprintf(cbuf,sizeof(cbuf),"circle # %d (%d)",rand(10000),i);
      parts[i]=(std::shared_ptr<Shape>(new Circle(cbuf,rand(100))));
    } else {
      snprintf(cbuf,sizeof(cbuf),"square # %d (%d)",rand(10000),i);      
      parts[i]=(std::shared_ptr<Shape>(new Square("shape # " +name,rand(100))));
    }
  }
}

public: void sort() {
  std::sort(parts.begin(), parts.end(),
            [](const std::shared_ptr <Shape> & a, const std::shared_ptr <Shape> & b)
            { return a->name > b->name; });
}

public: void draw() {
  for (auto shape : parts) {
    shape->draw();
  }
}
  
public: void run() {
  int n = 100; // 1000000;
  Shapes shapes;
  long t0,t1;

  t0=clock();
  shapes.make(n);
  t1=clock();
  std::cout << "construct: " << double(t1-t0)/CLOCKS_PER_SEC << "s" << std::endl;
  t0=clock();
  shapes.sort();
  t1=clock();
  std::cout << "sort: " << double(t1-t0)/CLOCKS_PER_SEC << "s" << std::endl;    
}


};

int main() {
  Shapes shapes;
  shapes.run();
  return 0;
}

File: bench/Shapes.java
import java.util.*;
import java.io.*;
import java.nio.charset.Charset;
import java.nio.file.*;
import javax.script.*;

class Config {
    public static final File thisDir;
    static {
	File _thisDir = null;
	try {
	    _thisDir = new File(Config.class.getProtectionDomain().getCodeSource().getLocation().getPath());
	} catch (Exception ex) {
	    ex.printStackTrace();
	}
	thisDir = _thisDir;
    }

    public final Map<String, Object> map;

    public Config(String filename) {
	Map<String,Object> _map = null;
	try {
	    String configFilename = thisDir + "/" + filename;
	    byte[] configBytes = Files.readAllBytes(Paths.get(configFilename));
	    String configString = new String(configBytes,
					     Charset.forName("UTF-8")).replaceAll("\\r?\\n"," ");
	    ScriptEngine engine = new ScriptEngineManager().getEngineByName("nashorn");
	    Object object = engine.eval(String.format("JSON.parse('%s')", configString));
	    _map = Collections.unmodifiableMap((Map<String, Object>) object);
	} catch (Exception ex) {
	    throw new RuntimeException(ex);
	}
	map=_map;
    }

    public static final Config CONFIG = new Config("config.json");

    public int getN() {
	return (Integer) map.get("n");
    }
}

class Shape {
    String name;
    Shape(String _name) { name=_name; }
    void draw() { System.out.println("drawing: " + name); }
}

class Circle extends Shape {
    double radius;
    Circle(String _name, int _radius) { super(_name); radius=_radius; }
    void draw() { System.out.println("drawing circle: " + name + " r=" + radius); }  
}

class Square extends Shape {
    double side;
    Square(String _name, int _side) { super(_name); side=_side; }
    void draw() { System.out.println("drawing circle: " + name + " s=" + side); }  
}

class Shapes extends Shape {
    Shapes() { super("shapes"); }
    Random random = new Random();
    int rand(int n) { return random.nextInt(n); }

    ArrayList < Shape > parts = new ArrayList<Shape>();
    void make(int n) {
	for (int i=0; i<n; ++i) {
	    name = "" + rand(10000)  + "(" + i + ")";
	    if ((i%10) <= 5) {
		parts.add(new Circle("circle # " + name,rand(100)));
	    } else {
		parts.add(new Square("shape # " +name,rand(100)));
	    }
	}
    }

    void sort() {
	parts.sort((a,b)->a.name.compareTo(b.name));
    }

    void draw() {
	for (Shape shape : parts) {
	    shape.draw();
	}
    }
  
    public static void main(String[] args) {
	new Shapes().run();
    }

    void run() {
        int n = Config.CONFIG.getN();
	System.out.println("n=" + n);
	Shapes shapes = new Shapes();
	long t0,t1;
      
	t0=System.currentTimeMillis();
	shapes.make(n);
	t1=System.currentTimeMillis();
	System.out.println("construct: " + (t1-t0)/1000.0 + "s");
	t0=System.currentTimeMillis();
	shapes.sort();
	t1=System.currentTimeMillis();
	System.out.println("sort: " + (t1-t0)/1000.0 + "s");
    }
}
